### ISSUE TITLE
found "no jep in java.library.path" error

### ISSUE THREAD

**Issue by mokundong** [polynote/polynote#521]
after fill in the conf&dependencies ,error come when i click the "save&restart" button

**Comment by jeremyrsmith** [comment]
Welcome! Thanks for the error report! Just a couple of questions so we can figure out how to solve this for anyone who's experiencing it.

How did you start Polynote? Do you have jep installed? Can you tell us which platform you're using?

Polynote's run script tries to find jep by using some script commands:

```
pip3 show jep | grep "^Location:" | cut -d ':' -f 2 | cut -d ' ' -f 2
```

What happens if you run that at the terminal?

**Comment by mokundong** [comment]
thanks for response, when i run this scrip came out this:
`
/Users/mkd/miniconda3/lib/python3.7/site-packages
`

**Comment by jeremyrsmith** [comment]
Thanks for the screenshot; that's really helpful.

A couple of issues I can see:
- `pyspark` has to be installed with `pip3` if you're going to use it; it doesn't work properly if it's a notebook dependency.
- The "Resolvers" section is for repositories, but it looks like you've listed the `pandas` package there. You'd want `pandas` in your dependencies (where you have `pyspark`). Unelss it's also a pip repository?

But I think this particular error is happening because `/Users/mkd/miniconda3/lib/python3.7/site-packages` isn't the right location for the jep library. So something's going wrong with that script.

Can you try starting polynote this way instead (from the directory where the `polynote` script and JAR are located)

```
export POLYNOTE_DEPS=($(ls deps/*.jar))
java -cp polynote.jar:$(IFS=: ; echo "${POLYNOTE_DEPS[*]}") -Djava.library.path=/Users/mkd/miniconda3/lib/python3.7/site-packages/jep polynote.Main
```

Basically you need `polynote.jar` and everything in `deps` on the classpath, and you need wherever `libjep.dylib` is to be on `java.library.path`. I'm assuming jep is in `/Users/mkd/miniconda3/lib/python3.7/site-packages/jep`.

We'll try to figure out why the script can't find jep properly. Lots of people are having trouble with our run script (it's just always worked for us) so any help improving it is welcome!

**Comment by sbrunk** [comment]
Note that in an Anaconda Python installation there might be no `pip3` command at all, just `pip` even for Python 3 causing the startup script to fail finding jep.

https://github.com/polynote/polynote/blob/a62eb8e036960fbc50d76c87c808a88ead422e7e/scripts/polynote#L5

**Comment by jonathanindig** [comment]
Thanks for the pointer @sbrunk! Hopefully #532 will work with Anaconda Python installations. 

It seemed to work for me when I tested with a [miniconda3 docker image ](https://hub.docker.com/r/continuumio/miniconda3).

**Comment by dpatchigolla** [comment]
I am facing similar issue executing Python cells. I tried the above fix (with path pointing to where my jep is located), but still doesn't work for me. I get the following error:

`<home>/anaconda3/lib/python3.6/site-packages/jep/libjep.jnilib: dlopen(<home>/anaconda3/lib/python3.6/site-packages/jep/libjep.jnilib, 1): Library not loaded: @rpath/libpython3.6m.dylib
Referenced from: <home>/anaconda3/lib/python3.6/site-packages/jep/libjep.jnilib
Reason: image not found`

The jep path is what I got by tying in the command 
`pip3 show jep | grep "^Location:" | cut -d ':' -f 2 | cut -d ' ' -f 2`

**Comment by Marioqwe** [comment]
@dpatchigolla 

You might need to tell `libjep.jnilib` where to look for `libpython3.6m.dylib`. Try the following:

`install_name_tool -add_rpath ~/anaconda3/lib ~/anaconda3/lib/python3.6/site-packages/jep/libjep.jnilib`.

**Comment by danielecook** [comment]
I am getting this error:

```
Uncaught exception: /Users/dancook/opt/anaconda3/lib/python3.7/site-packages/jep/libjep.jnilib: dlopen(/Users/dancook/opt/anaconda3/lib/python3.7/site-packages/jep/libjep.jnilib, 1): Library not loaded: @rpath/libpython3.7m.dylib Referenced from: /Users/dancook/opt/anaconda3/lib/python3.7/site-packages/jep/libjep.jnilib Reason: image not found (java.lang.UnsatisfiedLinkError)
```

**Comment by ahirtz** [comment]
Hi,
I found a solution for OS X:

> [link] : https://github.com/ninia/jep/wiki/FAQ#how-do-i-fix-unsatisfied-link-error-no-jep-in-javalibrarypath

```
cd ./polynote
conda create --name polynote
conda activate polynote
pip install jep
export DYLD_LIBRARY_PATH="<HOME>/anaconda3/lib/python3.7/site-packages/jep:$DYLD_LIBRARY_PATH"
export POLYNOTE_DEPS=($(ls deps/*.jar))
export JEP_PATH=$(pip show jep | grep "^Location:" | cut -d ':' -f 2 | cut -d ' ' -f 2)
java -cp polynote.jar:$(IFS=: ; echo "${POLYNOTE_DEPS[*]}") -Djava.library.path=$JEP_PATH/jep polynote.Main
```

**Comment by manuzhang** [comment]
Also from https://github.com/ninia/jep/wiki/FAQ#how-do-i-fix-unsatisfied-link-error-no-jep-in-javalibrarypath, setting `PYTHONHOME` works for me.

> New in version 3.8: Jep will try harder to find the native library for you if you set the PYTHONHOME environment variable. When using Anaconda, set PYTHONHOME to your Anaconda installation directory, e.g. $(HOME)/anaconda3.

**Comment by mokundong** [comment]
> java -cp polynote.jar:$(IFS=: ; echo "${POLYNOTE_DEPS[*]}") -Djava.library.path=/Users/mkd/miniconda3/lib/python3.7/site-packages/jep polynote.Main

I got error like this :
> Uncaught exception: /Users/mkd/miniconda3/lib/python3.7/site-packages/jep/libjep.jnilib: dlopen(/Users/mkd/miniconda3/lib/python3.7/site-packages/jep/libjep.jnilib, 1): Library not loaded: @rpath/libpython3.7m.dylib Referenced from: /Users/mkd/miniconda3/lib/python3.7/site-packages/jep/libjep.jnilib Reason: image not found (java.lang.UnsatisfiedLinkError)

**Comment by manuzhang** [comment]
@mokundong have you tried setting `PYTHONHOME` ?

**Comment by mokundong** [comment]
> @mokundong have you tried setting `PYTHONHOME` ?
> 
> ```
> export PYTHONHOME=/Users/mkd/miniconda3/
> ```

> yes. I give up this , I'm running fine with docker

**Comment by manuzhang** [comment]
I should mention after setting `PYTHONHOME`,  just `./polynote.py` will do. No need to manually set Java classpath
